xarray/core/variable.py

```diff
@@ -179,6 +179,9 @@ def as_compatible_data(data, fastpath=False):
         else:
             data = np.asarray(data)
 
+    if hasattr(data, '__array_function__'):
+        return data
```

Comment:
`numpy.ndarray` has a `__array_function__` attribute, so this won't quite do -- we still want to do the conversions on the lines below on NumPy arrays.

Also, it's at least possible to have a `__array_function__` attribute even if overrides with NEP-18 are not enabled (e.g., with NumPy version 1.16 without the environment variable set). We might want to raise an error in this case? Dask has a little utility for detecting this situation which might be worth borrowing: https://github.com/dask/dask/blob/c792364d1528c9c688a66c4d116e5f3b49776bb8/dask/array/utils.py#L352

Reply:
Sounds good. See 1b62f6d1. I added checks for ndarray-ness and NEP18.